cpp/src/barretenberg/ecc/groups/affine_element_impl.hpp

```diff
@@ -229,5 +229,38 @@ affine_element<Fq, Fr, T> affine_element<Fq, Fr, T>::hash_to_curve(const uint64_
 
     return affine_element<Fq, Fr, T>(x_out, y_out_);
 }
+
+template <typename Fq, typename Fr, typename T>
+affine_element<Fq, Fr, T> affine_element<Fq, Fr, T>::random_element(numeric::random::Engine* engine) noexcept
+{
+    if (engine == nullptr) {
+        engine = &numeric::random::get_engine();
+    }
+
+    bool found_one = false;
+    Fq yy;
+    Fq x;
+    Fq y;
+    while (!found_one) {
+        // Sample a random x-coordinate and check if it satisfies curve equation.
+        x = Fq::random_element(engine);
+        yy = x.sqr() * x + T::b;
+        if constexpr (T::has_a) {
+            yy += (x * T::a);
+        }
+        auto [found_root, y1] = yy.sqrt();
```

Comment:
When `!found_root` it makes sense to skip the work below here (sampling a random byte and doing a field negation), but doesn't matter unless we plan to generate a lot of points with this function.

Reply:
This was Kesha's suggestion and I agree with him. If we don't sample a bit and negate the y-coordinate, we're sampling a random point from only "upper" half of the curve points. To really get a "random" point on the curve, we need to randomly change the sign of the `y` coordinate.

Reply:
Yes, this make sense, but my point is that if `found_root` is `false`, then you don't need to bother sampling a random byte and negating a coordinate, so it would be more efficient to to wrap everything below line 251 in an `if (found_root)` block. About half the time the `while` loop will execute more than once, meaning that if you're generating 2 billion random points of the curve, then you will sample about 1 billion unneeded bytes and do about 1 billion unneeded 256-bit field element negations.